### ISSUE TITLE
`invalid-job-id` errors on `tProxy` don't result in rejection of shares

### ISSUE THREAD

**Issue by plebhash** [stratum-mining/stratum#791]
I was trying a BitAxe Ultra (BM1366) and I noticed multiple errors like this:
```
ERROR translator_sv2::lib::proxy::bridge: Submit share error Ok("invalid-job-id")
```

It also happens with `CPUminer`.

It was reported here: https://github.com/stratum-mining/stratum/issues/788

as [stated](https://github.com/stratum-mining/stratum/issues/788#issuecomment-1987870478) by @Fi3:
> This happens because the miner keep sending shares the for the old job and the translator mark them as invalid 

That explains the root cause of the error. Let's call these shares "delayed shares".

However, delayed shares are still accounted as "accepted" on the BitAxe side.

![image](https://github.com/stratum-mining/stratum/assets/147345153/5ba63f03-b2eb-4dce-9fa4-0b1865c0bcc1)

That is likely because the delayed shares only trigger an error event, but never really trigger some share rejection.

**Comment by plebhash** [comment]
I'm labelling this issue as `good first issue`. As instructions on how to dive into this:
1. You would need an ASIC. CPUmining will not provide enough hashpower to reproduce this issue reliably.
2. Before starting, follow SRI [`Getting Started` tutorial](https://stratumprotocol.org/getting-started/) and do some CPU mining anyways. That will help you get familiar with the process of deploying a local SRI setup.
3. Try to reproduce it with your ASIC.

**Comment by Fi3** [comment]
yep if I remember correctly the translator do not bother to send a response downstream for share error. I guess that this is mainly cause is supposed to be a proxy used by miners, so it is where metrics are collected and you use it as observable point. Said that returning errors for invalid shares would be a very nice thing to add. Also I don't know if the downstream need that message for stop sending shares and start working on the new job (don't think so but maybe is the case)

**Comment by GitGab19** [comment]
Translator should be able to communicate to the miners the refused share whenever it's refused from upstream or it's not valid. What translator should do is to send downstream the following string: {"id": job_id, "result": `false`, "error": null} in those cases.
Right now it always sends it with `"result": true`, I just checked.

@plebhash we can reinsert the label good-first-issue imo, what do you think?

**Comment by plebhash** [comment]
Cool, I'm putting the `good-first-issue` label back again.

We believe this is a good first issue because the fix is a relatively simple change, so it should provide some smooth intro for new contributors.

But it still requires that newcomers get familiar with SRI basics. A good preparation is to simply follow the `Getting Started` tutorial from https://stratumprotocol.org. It will help you deploy the different SRI roles and do some CPU mining on testnet.

A BitAxe is not a strict requirement for this issue, although it could be useful since the AxeOS UI provides a counter on accepted vs rejected shares, which was how I first identified this issue.

**Comment by adoerr** [comment]
I will give it a try. I'm going to start with working through the SRI tutorial and try to figure out how to hook up my BitAxe. This hopefully will allow me to reproduce the issue.

**Comment by adoerr** [comment]
>  What translator should do is to send downstream the following string: {"id": job_id, "result": false, "error": null} in those cases.
Right now it always sends it with "result": true, I just checked.

I think I'm a bit stuck in terms of finding a way how to actually change `result` to `true`. The issue is that [accepted](https://github.com/stratum-mining/stratum/blob/95d89b29d596b1ba9d1596e0da9bf206e19a0187/protocols/v1/src/lib.rs#L127-L128) is always `true` because  [handle_submit()](https://github.com/stratum-mining/stratum/blob/95d89b29d596b1ba9d1596e0da9bf206e19a0187/roles/translator/src/lib/downstream_sv1/downstream.rs#L493-L512)  always returns `true`.

[handle_submit_shares()](https://github.com/stratum-mining/stratum/blob/95d89b29d596b1ba9d1596e0da9bf206e19a0187/roles/translator/src/lib/proxy/bridge.rs#L237-L242) in turn only emits the `invalid-job-id` error message and doesn't return a useful error in this case either.

Any suggestions how to proceed without reworking error propagation itself  :thinking:

**Comment by Fi3** [comment]
share verification is not did at transloator::downstream level but at bridge level that is the one that have both sv2 and sv1 information (when you get a share if the share is above target you have to send it upstream). Downstream and bridge have differenet execution contexts. I guess that best thing to do, is to not send anything downs on share arrive as a transaltor, but send share responses from bridge to translator: https://github.com/stratum-mining/stratum/blob/95d89b29d596b1ba9d1596e0da9bf206e19a0187/roles/translator/src/lib/downstream_sv1/downstream.rs#L288
I think that you want to add a future to this select where you receive share responses

**Comment by 0xSaksham** [comment]
I want to give this a shot.

**Comment by adoerr** [comment]
> I want to give this a shot.

Please :+1:  I got completely buried by other stuff :see_no_evil:

**Comment by GitGab19** [comment]
> I want to give this a shot.

@0xSaksham are you working on this?
If so, I will assign it to you so that we avoid more people working on the same task

**Comment by GitGab19** [comment]
@0xSaksham are you working on this?

**Comment by 0xSaksham** [comment]
Hey, I'll need time. Please assign this to anyone else.

**Comment by plebhash** [comment]
@lorbax so when I suggested our informal MG workshop call, I think we could try to reproduce this issue as a guiding exercise

**Comment by lorbax** [comment]
> @lorbax so when I suggested our informal MG workshop call, I think we could try to reproduce this issue as a guiding exercise

sure DM me on discord

**Comment by plebhash** [comment]
I'm adding a MG test against this bug via https://github.com/stratum-mining/stratum/pull/910

currently, it is exprected that this test will fail, since the tProxy bug has not yet been fixed

on the PR that fixes this bug, we should also revert commit bb9c643ab598c895fd7959be5c70e0ad834bc280

**Comment by rockcoolsaint** [comment]
Anyone working on this?

**Comment by plebhash** [comment]
> > Anyone working on this?
> 
> Yeah I'm pretty sure @plebhash already fixed it here: https://github.com/stratum-mining/stratum/pull/910

correct

@rockcoolsaint we welcome reviews on #910

**Comment by rockcoolsaint** [comment]
> > > Anyone working on this?
> > 
> > 
> > Yeah I'm pretty sure @plebhash already fixed it here: #910
> 
> correct
> 
> @rockcoolsaint we welcome reviews on #910

Alright

**Comment by plebhash** [comment]
closed via #910
